Fix Pad op parsing

gyulaz-htec opened the pull request:
This PR fixes https://github.com/ROCm/AMDMIGraphX/issues/2763
The `constant_value` of the [Pad](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Pad) operator is parsed as an `undefined` instruction for the models in the issue.
In this case I think it's better to fall back to the default `0.0f` value than throwing an error.

umangyadav commented on src/onnx/parse_pad.cpp:
```diff
@@ -177,11 +177,14 @@ struct parse_pad : op_parser<parse_pad>
             }
             auto val_arg = val_ins->eval();
-            if(val_arg.get_shape().elements() != 1)
+            if(val_arg.get_shape().elements() > 1)
```
When there are more than 1 pad values, do they all have same value ?

gyulaz-htec replied:
Yes, the `const_value` is just a scalar value. If there are more than 1 pad values they all use the same `const_value`.

pfultz2 commented:
> The constant_value of the [Pad](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Pad) operator is parsed as an empty shape for the models in the issue.
In this case I think it's better to fall back to the default 0.0f value than throwing an error.

The `eval()` returns an empty shape when value arg can't be const folded. That means the constant value is dynamic in this case so I dont think its a good idea to ignore it and just set it to 0.

Also, can a unit test be added for this?

gyulaz-htec commented:
@pfultz2 the empty shapes comes from an `undefined` instruction in place of the `const_value` Pad input. Accroding to the [spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Pad) `const_vale` is optional, is it possible that this optional value is parsed as `undefined`?
I assumed that my guess is correct and updated the code that we handle the `undefined` there. Added a test case for it as well.

lakhinderwalia commented on src/onnx/parse_pad.cpp:
```diff
@@ -169,7 +169,7 @@ struct parse_pad : op_parser<parse_pad>
     {
         float value = 0.0f;
-        if(args.size() >= 3 and args.at(2)->get_shape().scalar())
+        if(args.size() >= 3 and args.at(2)->get_shape().scalar() and not args.at(2)->is_undefined())
```
The `if clause` (line 172) and the `if clause` (line 180) below: are they testing the same thing? Thanks.

gyulaz-htec replied:
No they don't.
An `undefined` shape has has `0` elements thus we want to avoid line 180 if we have `undefined`.
AFAIK scalar shape only checks if the shape is not dynamic or has non-zero stride.

lakhinderwalia replied:
Thanks. @pfultz2 I wonder if there should be just a single check for number of elements == 1. No need to check for scalar() here. And that should eliminate the check for is_undefined(), if it has zero elements.

lakhinderwalia replied:
@gyulaz-htec I don't want to hold your PR. . It is a bit illogical to first check for the shape of the argument, and then check the argument is not undefined -- the order should be reverse. But.. instead.. 
It is simpler to just check that the number of elements ==1, and just bypass the undefined (and scalar check) here. The error case when elements > 1 can be handled separately (or combined into elements >=1). 
I will approve the PR, but feel feel to include this suggestion if you think it is logical. Thanks.

gyulaz-htec replied:
Thanks @lakhinderwalia your suggestion seems more simpler than my solution. I've updated the code according to that.

lakhinderwalia replied:
Thanks @gyulaz-htec

gyulaz-htec commented on test/onnx/parse/pad_undef_const_val_test.cpp:
```diff
@@ -0,0 +1,38 @@
+    migraphx::program p;
+    auto* mm = p.get_main_module();
+    auto l0  = mm->add_parameter("0", migraphx::shape{migraphx::shape::float_type, {2, 2}});
+    mm->add_literal({migraphx::shape{migraphx::shape::int32_type, {4}}, {1, 1, 1, 1}});
+    mm->add_instruction(migraphx::make_op("undefined"));
```
This comes from pytorch: https://github.com/pytorch/pytorch/issues/84979
By default `value` is not set (`None`) for Pad. During torch export the `None` becamues an empty input.
That empty input becomes an `undefined` during parsing here: 
https://github.com/ROCm/AMDMIGraphX/blob/4b9d647041636a1ec54265b8a71686a0d0aac25f/src/onnx/onnx_parser.cpp#L400-L403

gyulaz-htec replied:
May I add a comment abot this issue with torch export?

umangyadav replied:
For example : 
https://github.com/ROCm/AMDMIGraphX/blob/4b9d647041636a1ec54265b8a71686a0d0aac25f/test/dead_code_elimination_test.cpp#L137

gyulaz-htec replied:
AFAIK DCE runs after the parsing, and this `undefined` is during parse.

umangyadav replied:
@pfultz2  do you know why "undefind" remains here despite running DCE ?

umangyadav replied:
> AFAIK DCE runs after the parsing, and this undefined is during parse. 

Yes after the parsing `optimize_onnx` would run DCE
https://github.com/ROCm/AMDMIGraphX/blob/4b9d647041636a1ec54265b8a71686a0d0aac25f/test/onnx/include/onnx_test.hpp#L47

gyulaz-htec replied:
DCE is not running because `run_passes` was not set for `optimize_onnx` in the test.
https://github.com/ROCm/AMDMIGraphX/blob/4b9d647041636a1ec54265b8a71686a0d0aac25f/test/onnx/include/onnx_test.hpp#L41-L49

umangyadav replied:
Oh. Right. I don't know what i was thinking.

lakhinderwalia reviewed (approved):
Left you a suggestion. Approving the PR.
